src/firecracker/build.rs

```diff
@@ -0,0 +1,96 @@
+
+    let seccomp_json_path = format!("{}/{}.json", JSON_DIR, target);
+    // If the current target doesn't have a default filter, use a default, empty filter.
+    // This is to make sure that Firecracker builds even with libc toolchains for which we don't
+    // provide a default filter. For example, GNU libc.
+    let seccomp_json_path = if Path::new(&seccomp_json_path).exists() {
+        seccomp_json_path
+    } else {
+        println!(
+            "cargo:warning=No default seccomp policy for target: {}. Defaulting to \
+             `resources/seccomp/unimplemented.json`.",
+            target
+        );
+        format!("{}/unimplemented.json", JSON_DIR)
+    };
+
+    // Retrigger the build script if the JSON file has changed.
+    // let json_path = json_path.to_str().expect("Invalid bytes");
+    println!("cargo:rerun-if-changed={}", &seccomp_json_path);
+    // Also retrigger the build script on any seccompiler source code change.
+    register_seccompiler_src_watchlist(SECCOMPILER_SRC_DIR);
+
+    let out_path = format!("{}/{}", out_dir, ADVANCED_BINARY_FILTER_FILE_NAME);
+
+    // Run seccompiler-bin, getting the default, advanced filter.
+    run_seccompiler_bin(&target_arch, &seccomp_json_path, &out_path);
+}
+
+// Run seccompiler with the given arguments.
+fn run_seccompiler_bin(arch: &str, input_path: &str, out_path: &str) {
+    let input = std::fs::read_to_string(input_path).expect("Correct input file");
+    let filters: JsonFile = serde_json::from_str(&input).expect("Input read");
+
+    let arch = arch.try_into().expect("Target");
+    let compiler = Compiler::new(arch);
+
+    // transform the IR into a Map of BPFPrograms
+    let bpf_data: BTreeMap<String, BpfProgram> = compiler
+        .compile_blob(filters.0, false)
+        .expect("Successfull compilation");
+
+    // serialize the BPF programs & output them to a file
+    let output_file = File::create(out_path).expect("Create seccompiler output path");
+    bincode::serialize_into(output_file, &bpf_data).expect("Seccompiler serialization");
+}
+
+// Recursively traverse the entire seccompiler source folder and trigger a re-run of this build
+// script on any modification of these files.
+fn register_seccompiler_src_watchlist<P: AsRef<Path>>(src_dir: P) {
+    let contents = std::fs::read_dir(src_dir).expect("Unable to read folder contents.");
+    for entry in contents {
+        let path = entry.unwrap().path();
+        let metadata = std::fs::metadata(&path).expect("Unable to read file/folder metadata.");
+
+        if metadata.is_file() {
+            // Watch all source files.
+            println!(
+                "cargo:rerun-if-changed={}",
+                path.to_str().expect("Invalid unicode bytes.")
+            );
+        } else if metadata.is_dir() {
+            // If is a folder, recurse.
+            register_seccompiler_src_watchlist(&path);
+        }
+    }
+}
```

Comment:
From the docs https://doc.rust-lang.org/cargo/reference/build-scripts.html#rerun-if-changed
> If the path points to a directory, it will scan the entire directory for any modifications.

Could we remove this and simplify line 51 to:

```rust 
            println!(
                "cargo:rerun-if-changed={}",
                SECCOMPILER_SRC_DIR.to_str().expect("Invalid unicode bytes.")
            );
```

Probably in a separate commit.

Reply:
good catch, will do

Reply:
done

Reply:
This was purposefully implemented like this because at that time, cargo was not looking at all of the sources recursively: https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorerun-if-changedpath.
The behaviour was OS-dependent and last time I checked, it did not detect changes in the filters unless I traversed every folder recursively.